### PR TITLE
reverse order in which stackplot elements are added to axes

### PR DESCRIPTION
**Pull request by pseyfert**
## PR Summary

This will does not alter the order in which stackplot elements are show
on the axes, colors are used, labels are assigned, etc. It is only
(meant to) change the order in which handles are stored in the axes
object and thus in which legend entries appear by default. With this
change the top most entry in the legend should be the top most plotted
stack plot component.

In an example (derived from https://matplotlib.org/3.1.1/gallery/lines_bars_and_markers/stackplot_demo.html)

```py
import numpy as np
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y1 = [1, 1, 2, 3, 5]
y2 = [0, 4, 2, 6, 8]
y3 = [1, 3, 5, 7, 9]

y4 = [23, 23, 23, 24, 25]

y = np.vstack([y1, y2, y3])

labels = ["Fibonacci ", "Evens", "Odds"]

fig, ax = plt.subplots()
ax.plot(x, y4, label="staying above")
ax.stackplot(x, y1, y2, y3, labels=labels)
ax.legend(loc='upper left')
plt.show()
```

In the example, the bottom most component of the stackplot is 'Fibonacci', yet by default 'Fibonacci' will be the top most entry in the legend. This reverse order causes mental overhead when reading a plot and becomes even more problematic for colorblind readers or once the color pallet is cycled through more than once.

I find the example in https://codeyarns.com/tech/2019-09-12-how-to-reverse-order-of-legend-in-matplotlib.html insufficient to address that inconvenience as the fix is needed for every usage and breaks in the example as I only want to affect the order of stackplot elements but the 'staying above' entry should remain where it is in the legend.

I enabled "Allow edits by maintainers".

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

- [ ] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).


#### pytest error

```
==================================================================================================================== test session starts ====================================================================================================================
platform linux -- Python 3.9.1+, pytest-6.0.2, py-1.10.0, pluggy-0.13.0
rootdir: /tmp/mpl/src/matplotlib, configfile: pytest.ini, testpaths: lib
plugins: flake8-1.0.6
collected 0 items / 1 error                                                                                                                                                                                                                                 

========================================================================================================================== ERRORS ===========================================================================================================================
_______________________________________________________________________________________________________________ ERROR collecting test session _______________________________________________________________________________________________________________
/usr/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1030: in _gcd_import
    ???
<frozen importlib._bootstrap>:1007: in _find_and_load
    ???
<frozen importlib._bootstrap>:972: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:228: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1030: in _gcd_import
    ???
<frozen importlib._bootstrap>:1007: in _find_and_load
    ???
<frozen importlib._bootstrap>:972: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:228: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1030: in _gcd_import
    ???
<frozen importlib._bootstrap>:1007: in _find_and_load
    ???
<frozen importlib._bootstrap>:986: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:680: in _load_unlocked
    ???
<frozen importlib._bootstrap_external>:790: in exec_module
    ???
<frozen importlib._bootstrap>:228: in _call_with_frames_removed
    ???
lib/matplotlib/__init__.py:107: in <module>
    from . import _api, cbook, docstring, rcsetup
lib/matplotlib/cbook/__init__.py:31: in <module>
    from matplotlib import _api, _c_internal_utils
E   ImportError: cannot import name '_c_internal_utils' from partially initialized module 'matplotlib' (most likely due to a circular import) (/tmp/mpl/src/matplotlib/lib/matplotlib/__init__.py)
================================================================================================================== short test summary info ==================================================================================================================
ERROR  - ImportError: cannot import name '_c_internal_utils' from partially initialized module 'matplotlib' (most likely due to a circular import) (/tmp/mpl/src/matplotlib/lib/matplotlib/__init__.py)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
===================================================================================================================== 1 error in 0.18s ======================================================================================================================
```

### CODE REVIEW

**Comment by jklymak** [comment]
This seems reasonable to me.  However, you will need to make the tests pass, and this is a breaking change (i.e. imagine someone has used the workaround, now their plot legend will be wrong) so we need to have some sort of deprecation path.

**Comment by pseyfert** [comment]
okay, eventually figured out how to run the tests. is there any established transition/deprecation path in matplotlib?

**Comment by jklymak** [comment]
I think you'll have to do a song and dance with a kwarg that default to the old behavior but warns, and then defaults to the new behavior?  Unless this is so obviously desirable that the old way was a bug, in which case we could maybe just have a kwarg to give the old behavior.

**Comment by pseyfert** [comment]
Sorry for the long silence.

I have an image to illustrate the original use case. In a stack plot like this, where colors get re-used, we were wondering "are the thick bands entries 1 and 2 or 9 and 10?". The plot is with my change and now in examples/lines_bars_and_markers/stackplot_demo.py.
stack_example

Talking to a few colleagues it's not clear to me that breaking the current behaviour is desired. While on the use case example, the updated ordering is appreciated, some also said the current behaviour would be obvious and expected. The name of the new option 'top_to_bottom' is up for discussion. It may seem clear from what it does internally (plot the areas top to bottom or bottom to top) but doesn't capture well what the observable change for a user would be.

(Going through the checklist from the template now)

**Comment by jklymak** [comment]
I'm popping to draft as this needs to pass the tests.  However, from the comment above, its not clear if this change is desired?

**Comment by pseyfert** [comment]
Imo the feature is desired, I'm bodging around it in production. If other users will appreciate turning it on-by-default or if off-by-default is a better setting is unclear to me.

**Comment by pseyfert** [comment]
okay, found the button to remove draft status.

let me just point out atm the branch is such that the default is unchanged.

**Review by tacaswell — DISMISSED** [review]
We could 🚲 🏠  the `bottom_area` and `not_bottom_areas` internal function names a bit, but I do not think it is worth holding over that.

**Comment by tacaswell** [comment]
@pseyfert Looks good!  I added / removed a bunch of label, I think they are now correct (no API change, has docs, has a test, and is a new feature).

**Review by timhoffm — COMMENTED** [review]
The implementation is good.

I'm not quite happy with the name `top_to_bottom`. It's unclear what this means. My first intuition on that name was, that for data `[y1, y2, y3]` it will plot y1 above y2 above y3.

Maybe something like `draw_top_first`? Name suggestions welcome.

File: lib/matplotlib/stackplot.py [hunk 1]
```diff
@@ -47,6 +47,14 @@ def stackplot(axes, x, *args,
         A sequence of labels to assign to each data series. If unspecified,
         then no labels will be applied to artists.
 
+    top_to_bottom : bool, optional
+        When true, the top most component of the stackplot will be plotted
+        first. This has the effect that the top most plot component will appear
+        above the other plot components in the legend. That is, for
+        ``stackplot(x, y1, y2, y3)``, the legend will contain y3 above y2 and
+        y2 above y1.
+        The default behaviour is plotting the bottom area first.
```
Comment (tacaswell) [inline lib/matplotlib/stackplot.py:56]:
```suggestion
        The default behaviour is plotting the bottom area first.
        
         .. versionadded:: 3.7
```

File: doc/users/next_whats_new/stackplot_legend.rst [hunk 2]
```diff
@@ -0,0 +1,24 @@
+Order of legend entries in stackplots
+-------------------------------------
+
+The order of entries in a legend of a `~matplotlib.axes.Axes.stackplot` can be inverted with the option ``top_to_bottom``. In the below example, a viewer would see the area for ``y3`` above ``y2``, and ``y2`` above ``y1``. Here the legend will be generated in the same way such that ``y3`` appears above ``y2`` and ``y2`` above ``y1``.
+
+.. code-block:: python
+
+   from matplotlib import pyplot as plt
+   plt.x = [1, 2, 3, 4, 5]
```
Comment (tacaswell) [inline doc/users/next_whats_new/stackplot_legend.rst:9]:
```suggestion
   x = [1, 2, 3, 4, 5]
```

File: doc/users/next_whats_new/stackplot_legend.rst [hunk 3]
```diff
@@ -0,0 +1,24 @@
+Order of legend entries in stackplots
+-------------------------------------
+
+The order of entries in a legend of a `~matplotlib.axes.Axes.stackplot` can be inverted with the option ``top_to_bottom``. In the below example, a viewer would see the area for ``y3`` above ``y2``, and ``y2`` above ``y1``. Here the legend will be generated in the same way such that ``y3`` appears above ``y2`` and ``y2`` above ``y1``.
+
+.. code-block:: python
```
Comment (tacaswell) [inline doc/users/next_whats_new/stackplot_legend.rst:6]:
```suggestion
.. plot:
```

**Comment by QuLogic** [comment]
> Maybe something like `draw_top_first`? Name suggestions welcome.

If it helps come up with names, it appears that the same problem occurs with stacked histograms. For example, if you add a legend to [the histogram with multiple data sets example](https://matplotlib.org/stable/gallery/statistics/histogram_multihist.html), then you get the following:
Figure_1

As you can see, the side-by-side plots appear to be correct (at least for LTR readings). Somehow the stack step is 'right', but stacked bar is reversed. We will likely want to fix these at the same time (though not necessarily the same PR), so a name that works for both would be good.

**Comment by timhoffm** [comment]
> Did a quick look and plotly and ggplot both put this kinda thing as an attribute on/method to legend, and they name it `traceorder` ('cause handles are traces in plotly) which is a string or `reverse` = True, and I wonder if something general like `label/legend/handle_order_reverse` would work?

Wow, great thinking! 🥇 We should definitively have that option on legends. It's not only solving the issue for all the above cases and e.g. `scatter()`, it's also more intuitive to place the functionality at the legend as the visible element affected rather than at the plotting function - while the plotting function (and more generally the order of commands) defines an initial order, it's almost natural that the legend should have the capability to rearrange the order.

**Comment by jklymak** [comment]
a) it would be nice if the draw order were consistent. b) I can understand where the OP is coming from in wanting the legend to be stacked the same way the histograms would be stacked, from top to bottom.  However, I would _personally_ expect the opposite, where the first entry of the legend to be the bottommost in the stack.  Perhaps more obviously, if the stacked bars were instead horizontal, I'd expect the top entry of the legend to correspond to the left-most element in the stack. 

Given that there is ambiguity in what people would expect, I'm personally somewhat in favour of this just being a legend keyword helper that basically does `handles=handles[::-1]; labels=labels[::-1]` for the user.  I'm particularly not in favour of changing the draw order given that I assume the same code is reused for horizontal stacking.

**Comment by timhoffm** [comment]
I think we have reached consensus that this should be realized via a legend parameter. I've opened a new issue for that: #24712 

@pseyfert Thanks for your contribution! Even though we did not take your implementation, it inspired us how we can make this a more general feature in the legend API.

**Pull request closed without merging**